EduVPN/AppDelegate.swift

```diff
@@ -99,35 +99,92 @@ class AppDelegate: NSObject, NSApplicationDelegate {
+             kAEShowRestartDialog, kAERestart:
+            return true
+        default:
+            return false
         }
+    }
+
+    private func showAlertConfirmingStopVPNAndQuit(
+        connectionService: ConnectionServiceProtocol) -> NSApplication.TerminateReply {
 
         let alert = NSAlert()
         alert.alertStyle = .warning
+
         alert.messageText = NSLocalizedString(
             "Are you sure you want to quit \(Config.shared.appName)?",
             comment: "macOS alert title on attempt to quit app")
         alert.informativeText = NSLocalizedString(
             "The active VPN connection will be stopped when you quit.",
             comment: "macOS alert text on attempt to quit app")
         alert.addButton(withTitle: NSLocalizedString(
-                         "Stop VPN & Quit",
-                         comment: "macOS alert button on attempt to quit app"))
+                            "Stop VPN & Quit",
+                            comment: "macOS alert button on attempt to quit app"))
+        alert.addButton(withTitle: NSLocalizedString(
+                            "Cancel", comment: "button title"))
+
+        func handleQuitConfirmationResult(_ result: NSApplication.ModalResponse) {
+            if case .alertFirstButtonReturn = result {
+                firstly {
+                    connectionService.disableVPN()
+                }.map { _ in
+                    NSApp.reply(toApplicationShouldTerminate: true)
+                }.cauterize()
+            } else {
+                NSApp.reply(toApplicationShouldTerminate: false)
+            }
+        }
+
+        if let window = NSApp.windows.first {
+            alert.beginSheetModal(for: window) { result in
+                handleQuitConfirmationResult(result)
+            }
+        } else {
+            let result = alert.runModal()
+            handleQuitConfirmationResult(result)
+        }
+
+        return .terminateLater
+    }
+
+    private func showAlertConfirmingLogoutWithVPNOn(
+        connectionService: ConnectionServiceProtocol) -> NSApplication.TerminateReply {
+
+        let alert = NSAlert()
+        alert.alertStyle = .warning
+
+        alert.messageText = NSLocalizedString(
+            "You are logging out while the VPN is on",
+            comment: "macOS alert title on attempt to quit app by logout")
+        alert.informativeText = NSLocalizedString(
+            "The VPN will be on when a user logs in next. You can turn off the VPN then by opening the \(Config.shared.appName) app.",
+            comment: "macOS alert text on attempt to quit app by logout")
+        alert.addButton(withTitle: NSLocalizedString(
+                            "Quit",
+                            comment: "macOS alert button on attempt to quit app by logout"))
         alert.addButton(withTitle: NSLocalizedString(
-                         "Cancel", comment: "button title"))
+                            "Stop VPN & Quit",
```

Comment:
While it makes sense to add this button, I'd say a Cancel button would be good to keep as well. Kinda odd to be in an app and forced to quit.

Reply:
This alert is shown only when the user is trying to logout/shutdown/restart. I don't think the user would want to abort the logout just because the VPN is on (it's not like a file is left unsaved).